New module: Add Jboss Cli module (web_infrastructure/jbosscli)

Wolfant opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Is a new module, connect ansible with jboss cli,
jboss-cli.sh is needed on remote host because remote connectios could cause errors when jboss
is running on heavy loads.


##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

JBoss-Cli module

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible --version
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```
and

```
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```

##### ADDITIONAL INFORMATION

This module help to exec masive jboss cli commands from ansible
  

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Wolfant committed:
Jboss Cli Module for Ansible 

more info https://github.com/Wolfant/ansible-jbossCli

Wolfant committed:
New Version

Wolfant committed:
Code Santy Test

Wolfant committed:
Sanity Test

Output on stderr from ansible-doc is considered an error.

Wolfant committed:
Change Shebang and duplicated author

Commit: Antonio Insuasti <antonio@insuasti.ec>

    Change Shebang and duplicated author
     On branch master
            modified:   jbosscli.py

Wolfant committed:
Delete copy

Wolfant committed:
Change Shebang on rigth file

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,215 @@
+    description:
+      - Show the JBoss Cli output, commonly in DMR
+notes:
+  - "jboss-cli.sh need to be runing on client host, and $JAVA_HOME/bin is needeth in Client $PATH"
```
s/needeth/needed/

Akasurde replied:
s/runing/running/

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,215 @@
+"""
+
+EXAMPLES = """
+# chage scan-interval value, on user wildfy instalation
```
s/instalation/installation/

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,215 @@
+    command: /subsystem=deployment-scanner/scanner=default:write-attribute(name=scan-interval,value=6000)
+    cli_path: /home/user/wildfly-10.1.0.Final/bin
+
+#  change ExampleDS datasource user-name, on 192.168.20.55:9990 default instalation
```
s/instalation/installation/

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,215 @@
+    command: /subsystem=datasources/data-source=ExampleDS:write-attribute(name=user-name,value=other)
+    server: 192.168.20.55:9990
+
+# Undeploy the hello world application on wildfy server
```
s/wildfy/wildfly/

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,215 @@
+        result['changed'] = True
+        result['stdout'] = out
+    if rc != 0:
+        module.fail_json(name='jbosscli', msg=re.findall("failure-description.+", out))
```
Here, you might want to return `out` if `re.findall`  returns `[]`

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,215 @@
+    required: false
+    default: /usr/share/wildfly/bin
+    description:
+      - The location in the filesystem where jboss-cli.sh is
```
missing 'located.' at end.

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,215 @@
+    required: false
+    default: localhost:9990
+    description:
+      - JBoss server or domain controller, whit management port
```
Possible typo.

sm4rk0 commented:
Thank you, @Wolfant
That's a valuable contribution.
Did you check what @liquidat did in https://github.com/ansible/ansible-modules-extras/pull/1862 ?
I'm using that module with some [local enhancements](https://gist.github.com/sm4rk0/276b1fdb529fab8d928b7c100fb4f0d8) (changed `command` with `commands` and improved return values).

Wolfant committed:
suggested changes

change typo and return

Wolfant commented:
@sm4rk0 i see yours and @liquidat, i try to change JBoss Cli output to json but fail when you use --file and the file have 2 or more commands, or when you have "undefined" data, check my with src instead of command

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+"""
+
+EXAMPLES = """
+# chage scan-interval value, on user wildfly installation
```
s/chage/change/

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+RETURN = '''
+---
+changed:
+    description: if jboss cli command change something
```
s/jboss/JBoss/

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+'''
+
+import os
+import shutil
```
Unused import

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+
+import os
+import shutil
+import time
```
Unused import

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+import shutil
+import time
+import re
+import grp
```
Unused import

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+    cli_path = module.params['cli_path']
+    server = module.params['server']
+    verbose = module.params['verbose']
+    jsout = None
```
Unused variable.

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+        ),
+    )
+
+    changed = False
```
Unused variable.

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+    verbose = module.params['verbose']
+    jsout = None
+
+    if command and src:
```
You can use `mutually_exclusive` for `command` and `src`. You can check [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L668) and [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/cloudformation.py#L410)

Wolfant replied:
@Akasurde thanks for your advice, and sorry for my English

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+    description:
+      - Show the JBoss Cli output, commonly in DMR
+notes:
+  - "jboss-cli.sh need to be running on client host, and $JAVA_HOME/bin is needed in Client $PATH"
```
s/Client/client/

Akasurde commented on lib/ansible/modules/web_infrastructure/jbosscli.py:
```diff
@@ -0,0 +1,219 @@
+        module.fail_json(msg="Argument 'src' and 'command' are mutually exclusive")
+
+    if user and not password:
+        module.fail_json(msg="Argument 'user' need 'password' ")
```
s/need/needs/

Wolfant committed:
Change Documentation

And use of mutually_exclusive
 On branch devel
 Changes to be committed:
	modified:   jbosscli.py

Wolfant committed:
Sanity

found 1 pep8 issue(s) which need to be resolved:
ERROR:
lib/ansible/modules/web_infrastructure/jbosscli.py:139:5: E303 too many
blank lines (2) (100%)
 On branch devel
 Changes to be committed:
	modified:   jbosscli.py

Wolfant commented:
ready_for_review

smbambling commented:
Was going to look at developing a module that used the jboss http api to do something similar.  Any reason you went with using the the jboss-cli.sh script vs hitting the http api end point locally on the host ?

wbrefvem commented:
@smbambling 
I have been thinking of doing something similar. I've also been looking at providing extra features for [Jolokia](https://jolokia.org/)-enabled instances and have started working on a Jolokia [python client](https://github.com/wbrefvem/python-jolokia) towards that end. My thought was that I would modify the existing jboss module.

smbambling commented:
@wbrefvem From a quick look it seems the current jboss module just drops a war/ear to be deployed.  If you do end up working on it and want a hand or someone to help test let me know.

Wolfant commented:
@wbrefvem  the existing jboss module just drop an artifact file and deploy it, do not permit really administration of JBoss.

Wolfant commented:
@smbambling  thanks, please test it,  It would be very helpful

wbrefvem commented:
@Wolfant 
@smbambling 
For whomever is interested, we're putting together a JBoss/Wildfly working group to coordinate work around Ansible & JBoss. We need at least two members to get it started. Email me or respond to the [thread](https://groups.google.com/forum/#!topic/ansible-devel/nttHHu0nMxI) on the ansible-devel list.

LeoUfimtsev commented:
would be nice to have this module available. We have a client who would like to use this functionality.

gundalow commented:
@LeoUfimtsev Hi, have you tested this module does it work for you?

rthaenert commented:
I am using this module currently for performing CLI commands on various JBoss instances (EAP 6 as well as EAP 7) and it is working fine.

One minor thing: The error handling is incorrect if `jboss-cli.sh` returns an error code:
```
cause = re.findall("Caused.+", err)
        if cause is not None:
```
re.findall() would return an empty list if there is no match which leads to incomplete error messages (for example if the JRE for the jboss cli is incorrect).

hvisage commented:
Progress on including this?

Easiest to "test" this locally?
